src/tactic/transport_2.lean

```diff
@@ -0,0 +1,141 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import tactic.equiv_rw
+
+namespace tactic
+open tactic.interactive
+
+/--
+Given `s : S α` for some structure `S` depending on a type `α`,
+and an equivalence `e : α ≃ β`,
+try to produce an `S β`,
+by transporting data and axioms across `e` using `equiv_rw`.
+-/
+meta def transport (s e : expr) : tactic unit :=
+do
+  (_, α, β) ← infer_type e >>= relation_lhs_rhs <|>
+    fail format!"second argument to `transport` was not an equivalence",
+  seq `[refine_struct { .. }]
+  (do
+    propagate_tags $ (do
+    f ← get_current_field,
+    mk_mapp f [α, none] >>= note f none,
+    b ← target >>= is_prop,
+    if b then (do
+      unfold_projs_target,
+      `[simp only [eq_rec_constant, eq_mpr_rfl, equiv.symm_apply_apply, equiv.arrow_congr'_apply, equiv.to_fun_as_coe]],
+      -- Explain/understand when this is/isn't appropriate, and don't just `try` it blindly.
+      try $ under_binders $ to_expr ``((%%e).symm.injective) >>= apply,
+      equiv_rw_hyp f e,
+      get_local f >>= exact) <|>
+      skip
+    else try (do
+      equiv_rw_hyp f e,
+      get_local f >>= exact)))
+
+namespace interactive
+open lean.parser
+open interactive interactive.types
+open tactic
+
+local postfix `?`:9001 := optional
+
+/--
+Given a goal `⊢ S β` for some parametrized structure type `S`,
+`transport` will look for a hypothesis `s : S α` and an equivalence `e : α ≃ β`,
+and attempt to close the goal by transporting `s` across the equivalence `e`.
+
+```lean
+example {α : Type} [ring α] {β : Type} (e : α ≃ β) : ring β :=
+by transport.
+```
+
+You can specify the object to transport using `transport s`,
+and the equivalence to transport across using `transport s with e`.
+-/
+meta def transport (s : parse texpr?) (e : parse $ (tk "with" *> ident)?) : itactic :=
+do
+  s ← match s with
+  | some s := to_expr s
+  | none := (do -- no object specified, try to find a hypothesis that looks like the target:
+    S ← target >>= (λ t, match t with
+    | expr.app S α := pure S
+    | _ := fail "No object to transport specified, and target doesn't look like a parametrized type."
+    end),
+    find_local ``(%%S _))
+  end,
+  e ← match e with
+  | some e := get_local e
+  | none := do -- no equivalence specified, try to find a hypothesis of the right shape:
+    (S, α) ← infer_type s >>= (λ t, match t with
+    | expr.app S α := pure (S, α)
+    | _ := fail format!"Object to transport doesn't look like a parametrized type: {s}"
+    end),
+    S ← whnf S,
+    β ← target >>= (λ t, match t with
+    | expr.app S' β := (do
+      S' ← whnf S',
+      (is_def_eq S' S >> pure β)
+        <|> (do S ← pp S, S' ← pp S', fail format!"Target doesn't match expected type: {S'} ≠ {S}"))
+    | _ := fail format!"Target doesn't match expected type: {S} _"
+    end),
+    find_local ``(%%α ≃ %%β)
+  end,
```

Comment:
I don't much like these match statements, trying to guess the intended arguments. They seem quite fragile (matching on `expr.app _ _`), but on the other hand it seems to work, and we could revise later.

The alternative is just to make both arguments mandatory. The only reason I wanted to avoid this is that often the `s` argument is a typeclass, and so might not have a good local name.

Reply:
I don't like these match statements either.  They would also fail with nested types, or if the types have multiple parameters.  Could `transport` work like `simpa`?  Then you would only need to specify the equivalence assumption.

For example:
```lean
example {α : Type} [ring α] {β : Type} (e : α ≃ β) : ring β :=
by transport e

-- which is equivalent to
example {α : Type} [h : ring α] {β : Type} (e : α ≃ β) : ring β :=
by equiv_rw e at h; exact h
```

You could first guess the equivalence, and then look for the argument whose type contains the lhs of the equivalence.

Reply:
I'll try it out tomorrow.

But --- that 'is equivalent to' is not how you should think of `transport`. `transport` is for the case when we don't already know how to rewrite under `ring` (so `equiv_rw` won't work!), and we're going to syntactically break apart our `ring α` and use `equiv_rw` to move its constituent parts over to build a `ring β`.

Reply:
Oh, I didn't realize that from the description.  So `transport` is then the first half of `@[derive equiv_functor]`?

In that case you could also find the assumption with the same `h.get_app_fn.const_name` (which returns `ring` in this case).

Reply:
Okay, I've removed most of the munge here. Now it's compulsory to specify the equivalence you want to transport along, and I guess the thing to transport by comparing with the target using `h.get_app_fn.const_name` as you suggested.